services/libs/aster-std/src/net/iface/util.rs

```diff
@@ -47,29 +49,27 @@ impl BindPortConfig {
 }
 
 pub fn spawn_background_poll_thread(iface: Arc<dyn Iface>) {
-    // FIXME: use timer or wait_timeout when timer is enable.
     let task_fn = move || {
-        debug!("spawn background poll thread");
+        trace!("spawn background poll thread for {}", iface.name());
+        let wait_queue = iface.polling_wait_queue();
         loop {
-            let next_poll_time = if let Some(next_poll_time) = iface.next_poll_at_ms() {
-                next_poll_time
-            } else {
-                Thread::yield_now();
+            let Some(next_poll_at_ms) = iface.next_poll_at_ms() else {
+                // Sleep for a rather long timeout
+                wait_queue.wait_timeout(&Duration::from_secs(10));
                 continue;
             };
-            let now = read_monotonic_milli_seconds();
-            if now > next_poll_time {
-                // FIXME: now is later than next poll time. This may cause problem.
+
+            let now_as_ms = read_monotonic_milli_seconds();
+
+            // FIXME: how to choose a better threshold?
+            const POLL_INTERVAL_THRESHOLD: u64 = 5;
+            if now_as_ms >= next_poll_at_ms - POLL_INTERVAL_THRESHOLD {
```

Comment:
I think we should remove this check as well as the nonsense constant `POLL_INTERVAL_THRESHOLD`.

If we call `iface.poll()` before `next_poll_at_ms`, `iface.poll()` will simply do nothing. In the second iteration, nothing changes, so we call `iface.poll()` again before `next_poll_at_ms`. This creates a spinning busy loop, and things are even worse now that kernel threads cannot be preempted.

I noticed this phenomenon months ago, and I think it hurts network performance, so I created a workaround when I did performance measurements in https://github.com/asterinas/asterinas/issues/485 (but I didn't open a PR because what I created is just a workaround). Anyway, I believe that this problem should be fixed, either in this PR or in a separate PR.

Reply:
Learned from [smoltcp doc](https://docs.rs/smoltcp/latest/smoltcp/iface/struct.Interface.html#method.poll_at),
> Return a soft deadline for calling [poll](https://docs.rs/smoltcp/latest/smoltcp/iface/struct.Interface.html#method.poll) the next time. The [Instant](https://docs.rs/smoltcp/latest/smoltcp/iface/struct.Instant.html) returned is the time at which you should call [poll](https://docs.rs/smoltcp/latest/smoltcp/iface/struct.Interface.html#method.poll) next. It is harmless (but wastes energy) to call it before the [Instant](https://docs.rs/smoltcp/latest/smoltcp/iface/struct.Instant.html), and potentially harmful (impacting quality of service) to call it after the [Instant](https://docs.rs/smoltcp/latest/smoltcp/iface/struct.Instant.html)

I assume (but cannot confirm) that `next_poll_at_ms` is relevant for the heartbeat packet or DHCP response, both of which may expire after a certain period of time. If we fail to conduct a poll prior to that time, it could potentially disrupt the connection. Therefore, I have decided to conduct a poll slightly in advance. However, I am uncertain if there is a more optimal solution available.

Reply:
Thanks for your explanation!

But if we can somehow call it before the `next_poll_at_ms`, we _must_ have a mechanism to avoid a spinning busy loop in case the `iface.poll()` actually does nothing (since the timer doesn't actually expire), right?

I think this sentence from the documentation you quote actually confirms my argument:
> It is harmless _(but wastes energy)_ to call it before the [Instant](https://docs.rs/smoltcp/latest/smoltcp/iface/struct.Instant.html)

We can of course call `iface.poll()` before `next_poll_at_ms`, but we have to behave normally (instead of putting ourselves in a busy loop) if `iface.poll()` cannot make any progress (so that `next_poll_at_ms` does not change in the second iteration).

As for:
> potentially harmful (impacting quality of service) to call it after the [Instant](https://docs.rs/smoltcp/latest/smoltcp/iface/struct.Instant.html)

I think this sentence is just warning us not to call `iface.poll()` too late, rather than suggesting that we call `iface.poll()` earlier than `next_poll_at_ms` (which is actually "wasting energy", as stated in the previous sentence). We can use thread priorities to achieve this goal (once thread priorities are properly implemented), not try to poll in advance.

Reply:
Yes, a spinning busy loop can significantly impact performance. I will fix it.